glossary.md

```diff
@@ -814,6 +814,129 @@ A custom allocator may be able to allocate memory faster, or use memory more eff
+
+```c++
+inline int get() {
+    return 1;
+}
+```
+
+This asks the compiler to [inline-expand](#inline-expands) your code.
+
+In addition the `inline` keyword has to be used for [definitions](#definition) in headers to not violate the [One-Definition Rule (ODR)](#one-definition rule):
+
+
+Learn more at <https://isocpp.org/wiki/faq/inline-functions#inline-member-fns>
+
+
+### one-definition rule
+
+The `inline` keyword has to be used for [definitions](#definition) in headers to not violate the One-Definition Rule (ODR):
+
+```c++
+// myheader.hpp
+
+inline int get() {
+    return 1;
+}
+
+template <typename T> int put(int x) {
+    return x + 1;
+}
+```
+
+The definition for `get` in the header needs `inline` - not for performance reasons - but to make sure multiple translation units do not clash exposing the same definition.
+
+Except for fully specialized templates (`template <>`) templates are `inline` by default and do not require the `inline` keyword when defining them in headers.
+It's a good practice to just always specify the `inline` keyword for definitions in headers.
+
+Learn more at <http://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule>
+
+### inline-expands
+
+Also know as "inlined", this describes when a compiler moves code such that "the function’s code gets inserted into the caller’s code stream" from <https://isocpp.org/wiki/faq/inline-functions#overview-inline-fns>
+
+Inlining functions allows the compiler's optimizer to "see" across [translation unit](#translation-unit) boundaries and optimize the code in its larger context.
+
+This expansion happens during compilation so the impact of inlining is on the resultant [object files](#object-files) and binaries produced. The compiler does not rewrite [source code](#source-code).
+
+However, let's use [source code](#source-code) as a simple example of what inlining is about.
+
+Given the following code:
+
+```c++
+inline int get() {
+    return 1;
+}
+
+int main() {
+    int result = get();
+    return result;
+}
+```
+
+The inline-expanded version would likely be seen/re-written internally by the compiler to be more like:
+
+```c++
+
+int main() {
+    int result = 1;
+    return result;
+}
+```
+
+What the compiler does is hard to predict (its job is to optimize code in innovative ways!), so this is an incomplete example, but hints at the start of what may be happening behind the scenes.
+You can use Clang's [`-Rpass`](https://clang.llvm.org/docs/UsersManual.html#options-to-emit-optimization-reports) switch to inspect what the inliner is doing and especially where it fails to inline.
+
+You could also imagine the compiler taking this example a step further and doing:
+
+```c++
+int main() {
+    return 1;
+}
+```
+
+Because in our example it knows that `result` will always be `1`.
+
+### inlined
+
+When a compiler [inline-expands](#inline-expands) a function.
+
+### definition
```

Comment:
Ah just came across this while writing the odr section: we should split `definition` and `declaration`.

    int get();

is a declaration, while

    int get() { return 1; }

is a declaration.

http://en.cppreference.com/w/cpp/language/declarations
http://en.cppreference.com/w/cpp/language/definition

Reply:
👍  thanks. 79560afc9cc2fcb13 looks great, please merge once you've had a chance to fix this. Thanks!

Reply:
rebased, added declaration vs definition, merged

Reply:
🎉  Thanks @daniel-j-h